File: PassDataTableToStoredProcedure/PassDataTableToStoredProcedureVB/Form1.vb
Imports System.Data.SqlClient

Public Class FormMain
    Private Sub BtnClear_Click(sender As Object, e As EventArgs) Handles BtnClear.Click

        TxtCustomer.Text = ""
        DtgClients.DataSource = Nothing

    End Sub

    Private Sub BtnSearch_Click(sender As Object, e As EventArgs) Handles BtnSearch.Click

        GetData()

    End Sub

    Private Sub TxtCustomer_KeyDown(sender As Object, e As KeyEventArgs) Handles TxtCustomer.KeyDown

        If (e.KeyCode = Keys.Enter) And (TxtCustomer.Text.Trim <> "") Then
            GetData()
        End If

    End Sub

    Private Sub GetData()
        Dim TmpTable As New DataTable
        Dim TmpTableRead As New DataTable
        Dim TmpRow As DataRow

        TmpTable.Columns.Add("CustomerId", GetType(String))
        TmpTable.Columns.Add("CustomerName", GetType(String))

        TmpRow = TmpTable.NewRow()

        TmpRow.Item("CustomerId") = TxtCustomer.Text
        TmpRow.Item("CustomerName") = "Nombre de Prueba"

        TmpTable.Rows.Add(TmpRow)

        Using TmpConnection As New SqlConnection("Data Source=.\sqlexpress01;Initial Catalog=Northwind;Persist Security Info=True;User ID=dbuser;Password=P4$$w0rd")

            TmpConnection.Open()

            Dim TmpSqlCommand As SqlCommand = New SqlCommand("dbo.SP_CustomerList", TmpConnection)
            TmpSqlCommand.CommandType = CommandType.StoredProcedure

            Dim TmpSqlParameter As SqlParameter = TmpSqlCommand.Parameters.AddWithValue("@CustomerList", TmpTable)

            TmpSqlParameter.SqlDbType = SqlDbType.Structured
            'TmpSqlCommand.ExecuteNonQuery()
            Dim TmpSqlDataAdapter As New SqlDataAdapter()

            TmpSqlDataAdapter.SelectCommand = TmpSqlCommand
            TmpSqlDataAdapter.Fill(TmpTableRead)

            If (TmpTableRead.Rows.Count > 0) Then

                DtgClients.DataSource = TmpTableRead
            Else
                MessageBox.Show("Code not found")
                DtgClients.DataSource = Nothing
            End If

        End Using

    End Sub

End Class

File: PassDataTableToStoredProcedure/PassDataTableToStoredProcedureCS/Form1.cs

using System;
using System.Data;
using System.Data.SqlClient;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace PassDataTableToStoredProcedureCS
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void TxtCustomer_KeyDown(object sender, KeyEventArgs e)
        {

            if( (e.KeyCode == Keys.Return) && (TxtCustomer.Text.Trim() != "") )
                GetData();
        }

        private void BtnSearch_Click(object sender, EventArgs e)
        {
            GetData();
        }

        private void BtnClear_Click(object sender, System.EventArgs e)
        {
            TxtCustomer.Text = "";
            DtgClients.DataSource = null;
        }

        private void GetData()
        {
            DataTable TmpTable = new DataTable();
            DataTable TmpTableRead = new DataTable();
            DataRow TmpRow;

            TmpTable.Columns.Add("CustomerId", typeof(string));
            TmpTable.Columns.Add("CustomerName", typeof(string));

            TmpRow = TmpTable.NewRow();

            TmpRow["CustomerId"] = TxtCustomer.Text.Trim();
            TmpRow["CustomerName"] = "Nombre de Prueba";

            TmpTable.Rows.Add(TmpRow);

            //SqlConnection TmpConnection;
            using (SqlConnection TmpConnection = new SqlConnection(@"Data Source=.\sqlexpress01;Initial Catalog=Northwind;Persist Security Info=True;User ID=dbuser;Password=P4$$w0rd" ))
            {

                TmpConnection.Open();

                SqlCommand TmpSqlCommand = new SqlCommand("dbo.SP_CustomerList", TmpConnection);
                TmpSqlCommand.CommandType = CommandType.StoredProcedure;

                SqlParameter TmpSqlParameter = TmpSqlCommand.Parameters.AddWithValue("@CustomerList", TmpTable);

                TmpSqlParameter.SqlDbType = SqlDbType.Structured;
                //TmpSqlCommand.ExecuteNonQuery();
                SqlDataAdapter TmpSqlDataAdapter = new SqlDataAdapter();

                TmpSqlDataAdapter.SelectCommand = TmpSqlCommand;
                TmpSqlDataAdapter.Fill(TmpTableRead);

                if (TmpTableRead.Rows.Count > 0)
                    DtgClients.DataSource = TmpTableRead;
               else
                    {
                    MessageBox.Show("Code not found");
                    DtgClients.DataSource = null;
                }

            }

        }

    }
}
